https://adventofcode.com/2022/day/8

In [138]:
import numpy as np
a= np.array([[1, 2, 3, 4,13], [5, 6, 7, 8,14], [9, 10, 11, 12,15],[16,17,18,19,20]])
a

array([[ 1,  2,  3,  4, 13],
       [ 5,  6,  7,  8, 14],
       [ 9, 10, 11, 12, 15],
       [16, 17, 18, 19, 20]])

In [28]:
c1 = a[:,0] # all column values in row value 0 --> column of values
r1 = a[1,:] # all row values in with column value 1 --> row of values
print(c1,r1)

[ 1  5  9 16] [ 5  6  7  8 14]


In [15]:
for n in range(1,len(c1)-1): # go through all values in the row/col
    tree = c1[n]
    left = c1[:n]
    right = c1[n+1:]
    print(f"{left}|{tree}|{right}")
    if tree < left.max():
        print("hidden from left")
    if tree < right.max():
        print('hidden from right')
    

[1]|5|[ 9 16]
hidden from right
[1 5]|9|[16]
hidden from right


OK, here's the strategy:
1. read in the data, create an np.matrix
1. create a second matrix of the same size, all TRUE
1. go thru the matrix, creating arrays for each row and column
1. for each value in the maxtrix, see if it is hidden from left and right (=up/down)
1. if hidden, change TRUE to FALSE (Visible?)
1. count N False

In [146]:
test = '''30373
25512
65332
33549
35390'''


In [147]:
input =[]
for line in test.split('\n'):
    line_lst = []
    for n in line:
        line_lst.append(int(n))
    input.append(line_lst)
input

[[3, 0, 3, 7, 3],
 [2, 5, 5, 1, 2],
 [6, 5, 3, 3, 2],
 [3, 3, 5, 4, 9],
 [3, 5, 3, 9, 0]]

In [148]:
a=np.array(input)
a

array([[3, 0, 3, 7, 3],
       [2, 5, 5, 1, 2],
       [6, 5, 3, 3, 2],
       [3, 3, 5, 4, 9],
       [3, 5, 3, 9, 0]])

In [30]:
print(a.shape)


(5, 5)


In [75]:
b=np.zeros(a.shape,dtype = int)
b

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])

In [16]:
b[1,3] = 1
b

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])

In [33]:
b[0,:] = 1
b[4,:] = 1
b[:,0] = 1
b[:,4] = 1
b

array([[1, 1, 1, 1, 1],
       [1, 0, 0, 0, 1],
       [1, 0, 0, 0, 1],
       [1, 0, 0, 0, 1],
       [1, 1, 1, 1, 1]])

In [ ]:
maxRow, maxCol = a.shape
for i in range(1,maxRow):
    testRow = a[i,:]
    for j in range(1,len(testRow)-1):
        tree = testRow[j]
        left = testRow[:j]
        right = testRow[j+1:]
        print(f"({i},{j})  |  {left.max()} |{tree}| {right.max()}")
        if tree > left.max() or tree > right.max():
            b[i,j] = 1
b

In [55]:
maxRow, maxCol = a.shape
for i in range(1,maxCol):
    testRow = a[:,i]
    for j in range(1,len(testRow)-1):
        tree = testRow[j]
        left = testRow[:j]
        right = testRow[j+1:]
        if tree > left.max() or tree > right.max():
            b[i,j] = 1
b

array([[1, 1, 1, 1, 1],
       [1, 1, 1, 0, 1],
       [1, 1, 0, 1, 1],
       [1, 0, 1, 0, 1],
       [1, 1, 1, 1, 1]])

In [64]:
a=np.array(input)
b=np.zeros(a.shape,dtype = int)
maxRow, maxCol = a.shape
b[0,:], b[maxRow-1,:],b[:,0], b[:,maxCol-1] = 1, 1, 1, 1 # edges are visible (=1)

for col in range(1,maxCol-1):
    for row in range(1,maxRow-1):
        test_row = a[:,row] 
        test_col = a[col,:]
        tree = a[col,row]
        above = test_col[:row]
        below = test_col[row+1:]
        left = test_row[:col]
        right = test_row[col+1:]
        #print(f"({col},{row}) T:{tree} A:{above} B:{below} L:{left} R:{right}")
        if tree > above.max() or tree.max() > below.max() or tree > left.max() or tree > right.max():
            b[col,row] = 1
            #print('visible')
        #else:
            #b[col,row] = 0
            #print('not visible')

np.count_nonzero(b)

1560

In [55]:
b

array([[0, 0, 0, 0, 0],
       [0, 1, 1, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 0, 0, 0]])

In [56]:
np.count_nonzero(b)

21

OK, the test works. Now let's try it with the actual data. . . 

In [157]:
wdir = '/home/alp/Google Drive/Python/AdventCode/2022'
with open (f"{wdir}/08input.txt") as f_in:
    data = f_in.read().split('\n')

input =[]
for line in data:
    line_lst = []
    for n in line:
        line_lst.append(int(n))
    input.append(line_lst)
#input

In [26]:
import numpy as np

a=np.array(input)
b=np.zeros(a.shape,dtype = int)
maxRow, maxCol = a.shape
b[0,:], b[maxRow-1,:],b[:,0], b[:,maxCol-1] = 1, 1, 1, 1 # edges are visible (=1)
# check rows:
for row_num in range(1,maxRow): # i = row number
    test_array = a[row_num,:]
    for col_num in range(1,len(testRow)-1): # j = column number
        tree = test_array[j]
        left = test_array[:j]
        right = test_array[j+1:]
        if tree > left.max() or tree > right.max(): # if taller than trees to left or right, it's visible (=1)
            b[col_num,row_num] = 1

# check cols:
for col_num in range(1,maxCol): # i = col number
    test_array = a[:,col_num]
    for row_num in range(1,len(testRow)-1): # j = row number
        tree = test_array[j]
        left = test_array[:j]
        right = test_array[j+1:]
        if tree > left.max() or tree > right.max():
            b[col_num,row_num] = 1

np.count_nonzero(b)

6945

Solutions:
Submitted 1132 (wrong)
  I realized that I wasn't referencing the B matrix correctly (a,b) = (row,col). 
  
  (WTF would they do this when all math grid systems are x,y == col,row?)
  
  So in the #check rows routine, the b matrix needs to be referenced as b[j,i], but in the #check cols, it needs to be b[i,j]
Submitted 1801 (wrong)
  This value is too high. 

How numpy addresses columns and rows is confusing me.
I rewrote the code to use one loop and better variable names, to help make things clearer.
And I got the right value:
  1560

In [160]:
a=np.array(input)
b=np.zeros(a.shape,dtype = int)
maxRow, maxCol = a.shape
b[0,:], b[maxRow-1,:],b[:,0], b[:,maxCol-1] = 1, 1, 1, 1 # edges are visible (=1)

for col in range(1,maxCol-1):
    for row in range(1,maxRow-1):
        test_row = a[row,:] 
        test_col = a[:,col]
        tree = a[row,col]
        
        above = test_col[:row]
        below = test_col[row+1:]
        left = test_row[:col]
        right = test_row[col+1:]
        #print(f"({col},{row}) T:{tree} A:{above} B:{below} L:{left} R:{right}")
        if tree > above.max() or tree.max() > below.max() or tree > left.max() or tree > right.max():
            b[row,col] = 1
            #print('visible')
        #else:
            #b[col,row] = 0
            #print('not visible')

np.count_nonzero(b)

1560

Part two: for a tree, how far is it to the edge of the map or to a tree of equal or greater height?

1. Same input and data matrix
1. Basic algorythm is the same (look at each tree).
1. test conditions are different: from tree, how far is it to a tree of equal or greater height?

The trick here is that
1. we need to reverse the lists when looking up and left, since we need to think about what can be seen from the tree
1. We need to create a function that counts how far it is to the next tree of equal height

In [151]:
col = 2
row = 3
#test_row = a[:,row] 
#test_col = a[col,:]
test_row = a[row,:] 
test_col = a[:,col]
tree = a[row,col]
above = test_col[row-1::-1]
below = test_col[row+1:]
left = test_row[col-1::-1]
right = test_row[col+1:]
print(f"({col},{row}) T:{tree} A:{above} B:{below} L:{left} R:{right}")

(2,3) T:5 A:[3 5 3] B:[3] L:[3 3] R:[4 9]


In [139]:
a

array([[ 1,  2,  3,  4, 13],
       [ 5,  6,  7,  8, 14],
       [ 9, 10, 11, 12, 15],
       [16, 17, 18, 19, 20]])

In [144]:
def dist_of_view(tree,view):
    for i in range(len(view)):
        if tree <= view[i]:
            #print(i)
            return i+1
    return i+1

In [152]:
views = [above,below,left,right]
product = 1
for view in views:
    print(view,dist_of_view(tree,view))
    product *= dist_of_view(tree,view)

print(product)

[3 5 3] 2
[3] 1
[3 3] 2
[4 9] 2
8


OK, now I've got the syntax down and the testing working.

(Again, WTF is the address (row,col)? Stupid)

In [158]:
a=np.array(input)
b=np.zeros(a.shape,dtype = int)
maxRow, maxCol = a.shape
b[0,:], b[maxRow-1,:],b[:,0], b[:,maxCol-1] = 1, 1, 1, 1 # edges are visible (=1)

def dist_of_view(tree,view):
    for i in range(len(view)):
        if tree <= view[i]:
            #print(i)
            return i+1 # b/c indexing is zero based!
    return i+1 # this is for the edges

for col in range(1,maxCol-1):
    for row in range(1,maxRow-1):
        test_row = a[row,:] 
        test_col = a[:,col]
        tree = a[row,col]
        
        above = test_col[row-1::-1]
        below = test_col[row+1:]
        left = test_row[col-1::-1]
        right = test_row[col+1:]
        
        views = [above,below,left,right]
        product = 1
        for view in views:
            #print(view,dist_of_view(tree,view))
            product *= dist_of_view(tree,view)
        b[row,col] = product

b.max()

252000

Go it correct on first attempt!